emcc.py

```diff
@@ -1431,7 +1437,7 @@ def default_setting(name, new_default):
   if settings.MAIN_MODULE or settings.SIDE_MODULE:
     settings.RELOCATABLE = 1
 
-  if settings.USE_PTHREADS and '-pthread' not in newargs:
+  if (settings.USE_PTHREADS or settings.WASM_WORKERS) and '-pthread' not in newargs:
```

Comment:
> (This could all go in an initial NFC PR that just adds SHARED_MEMORY, as I've been saying.)

This PR has been an NFC for one year and four months now. I hope not to split up this PR for further NFC PRs, because I don't think it is a split at any functional feature boundary. In other words, I don't think `-sSHARED_MEMORY` flag on its own has value in people targeting it on its own - and we certainly do not have any users that would be requesting that.

Reply:
> There was such a use case

Righto.. well I can move the `-sSHARED_MEMORY=1` from settings_internal.js back to settings.js and then you will get that shared memory and atomics and nothing else. There is nothing preventing it, except that the setting was marked in the internal file just for organization.

Although like I mentioned before, it is unclear what the practical end user programming interface with that would look like, and they will likely end up hacking the preamble/postamble/shell layer to customize it to their own uses.

> As a separately-useful feature it would be nice to land that separately.
> this is a massive PR

The largest parts here are the large test suite and the contents of library_wasm_workers.cpp/.js, which are quite internal to the feature and likely reviewing through all the individual test files is not that relevant (except for maybe the small test description comments that exist before each test in tests/test_browser.py)

> It is more work to split parts out

This is what I am mostly worried about. It will be a lot of work to do such PR refactor, but the desired end state of the repository is still the same as it already is in this PR. Also I am worried about getting stringed in to add more end user features: the `-sSHARED_MEMORY` build feature can get hairy, I am not sure if people know what they are asking for (they'll likely end up reimplementing on their own what WASM_WORKERS already implements).

> it also can lead to things landing faster overall in my experience because there is less back and forth.

I don't want to downplay the effort it takes to review this, but I still think that this should be manageable.

Splitting this PR to "system parts" vs "user library parts" of this PR should not change much, since it should be visible  from the file paths already in this PR to know which kind of change it is.

Or splitting this on SHARED_MEMORY vs WASM_WORKERS.. I really wish not to pursue the SHARED_MEMORY feature now. if I start doing a separate `-sSHARED_MEMORY` feature PR, it is then it becomes its own round of tests + docs + CI iteration, and there is even a question of gathering use cases on what kind of APIs people would want to use for that to provide for transfering memories and possibly modules, so that relevant things are tested. Waiting on the CI is close to two hours, and even small changes easily can drag on to two or more days on round tripping.

There should be nothing preventing such a user-facing SHARED_MEMORY build flag from being offered in the future even after this lands, and/or I can move the flag now from settings_internal.js to settings.js to already now allow that flag for advanced users, but I don't think I am ready now to start developing and testing that as a user facing feature.

Reply:
The -sSHARED_MEMORY path did land in its own PR as #16419, so this PR now only contains the WASM_WORKERS related code.

The -sSHARED_MEMORY feature is somewhat for advanced users only, there is only one rudimentary test in the suite that checks that the Memory is shared.. everything else is left to the developer in that mode.